src/operator/communication/shard.rs

```diff
@@ -0,0 +1,284 @@
+    /// Collect all shards of a stream at the same worker.
+    ///
+    /// The output stream in `receiver_worker` will contain a union of all
+    /// input batches across all workers.  The output streams in all other
+    /// workers will contain empty batches.
+    pub fn gather(&self, receiver_worker: usize) -> Stream<Circuit<P>, IB>
+    where
+        IB: Batch + Send,
+    {
+        match Runtime::runtime() {
+            None => self.clone(),
+            Some(runtime) => {
+                let num_workers = runtime.num_workers();
+                assert!(receiver_worker < num_workers);
+
+                if num_workers == 1 {
+                    self.clone()
+                } else {
+                    self.circuit()
+                        .cache_get_or_insert_with(
+                            GatherId::new((self.origin_node_id().clone(), receiver_worker)),
+                            move || {
+                                let (sender, receiver) = self.circuit().new_exchange_operators(
+                                    &runtime,
+                                    Runtime::worker_index(),
+                                    move |batch: IB, batches: &mut Vec<IB>| {
+                                        for _ in 0..num_workers {
+                                            batches.push(IB::empty(()));
+                                        }
+                                        batches[receiver_worker] = batch;
+                                    },
+                                    |trace: &mut Spine<IB>, batch: IB| trace.insert(batch),
+                                );
+                                // Is `consolidate` always necessary? Some (all?) consumers may be
+                                // happy working with traces.
+                                self.circuit()
+                                    .add_exchange(sender, receiver, self)
+                                    .consolidate()
+                            },
+                        )
+                        .clone()
+                }
+            }
+        }
+    }
+
+    // Partitions the batch into `nshards` partitions based on the hash of the key.
+    fn shard_batch<OB>(
+        batch: &IB,
+        nshards: usize,
+        builders: &mut Vec<OB::Builder>,
+        outputs: &mut Vec<OB>,
+    ) where
+        OB: Batch<Key = IB::Key, Val = IB::Val, Time = (), R = IB::R>,
+    {
+        builders.clear();
+
+        for _ in 0..nshards {
+            // We iterate over tuples in the batch in order; hence tuples added
+            // to each shard are also ordered, so we can use the more efficient
+            // `Builder` API (instead of `Batcher`) to construct output batches.
+            builders.push(OB::Builder::with_capacity((), batch.len() / nshards));
+        }
+
+        let mut cursor = batch.cursor();
+
+        while cursor.key_valid() {
+            let batch_index = (hash32(cursor.key()) as usize) % nshards;
```

Comment:
any reason to have the hash function be a parameter?

Reply:
You mean to _not_ be a parameter, so it can be supplied? Definitely, but probably can wait for a use case.